runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -391,12 +388,22 @@ function TSHighlighter._on_win(_, _win, buf, topline, botline)
   if not self then
     return false
   end
-  self.tree:parse({ topline, botline + 1 })
-  self:prepare_highlight_states(topline, botline + 1)
-  self.redraw_count = self.redraw_count + 1
+  local range = { topline, botline + 1 }
```

Comment:
Is it possible for this parse to change the tree outside the provided range (e.g. a full reparse, combined injections)?

In that case, If the user scrolls up/down before the parse finishes, these newly visible lines would need to be redrawn as well when it does finish.

Reply:
Though this does happen if there are 2 windows:

![image](https://github.com/user-attachments/assets/1dc6e484-7a61-4bfd-9f2b-80f513c07520)

Reply:
It seems that the parser becomes desynchronized from the buffer
![image](https://github.com/user-attachments/assets/612ccf5f-625d-4d0f-8560-33f18e10d047)

Reply:
Ah, thanks so much for the thorough testing. I can't find how to replicate this, do you mind if I ask what you did for the buffer split one? That one looks bad lol

Reply:
test.lua:
```
local _ = "
vim.api.nvim_set_hl(0, 'Test2', { bg = 'green' }) --repeated until line 50'000
vim.api.nvim_set_hl(0, 'Test3', { bg = 'blue' })
```

1. Open a split window.
2. Move the cursor to `"`
3. Enter insert mode and hold `"`, then release for a short period, then hold again

Reply:
For the two windows case I have noticed that after a few seconds for another reparse, the windows again come into sync with their highlighting despite the initial loss of sync due to parser lag. Are you able to confirm this happening?

As for the `:InspectTree` weirdness my guess is maybe that since the parser there is parsing synchronously it interferes/is different from the main highlighter parsing which runs async?

Reply:
> For the two windows case I have noticed that after a few seconds for another reparse, the windows again come into sync with their highlighting despite the initial loss of sync due to parser lag. Are you able to confirm this happening?
> 
> As for the `:InspectTree` weirdness my guess is maybe that since the parser there is parsing synchronously it interferes/is different from the main highlighter parsing which runs async?

Yes, they almost always synchronize if not many actions are performed. I typed and removed the `"` quickly and this happened. 

I assume both are essentially the same issue. When multiple parses are requested and the buffer is modified, sometimes the parser state becomes desynchronized.

In the `:InspectTree` case, I opened the preview to show that the root chunk is broken. The menu was opened after the fact.

Reply:
Thanks, that makes sense. I am realizing now how nice it would be to have first class async support in nvim

Reply:
In your opinion, should this issues block this PR? I'm unaware how much of this is just a natural result of using async functionality vs true edge cases I still need to account for